stan/math/rev/mat/functor/map_rect_reduce.hpp

```diff
@@ -0,0 +1,161 @@
+#ifndef STAN_MATH_REV_MAT_FUNCTOR_MAP_RECT_REDUCE_HPP
+#define STAN_MATH_REV_MAT_FUNCTOR_MAP_RECT_REDUCE_HPP
+
+#include <stan/math/prim/mat/functor/map_rect_reduce.hpp>
+#include <stan/math/prim/mat/fun/typedefs.hpp>
+#include <stan/math/rev/core/var.hpp>
+#include <stan/math/rev/mat/fun/typedefs.hpp>
+
+#include <vector>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+template <typename F>
+struct map_rect_reduce<F, var, var> {
+  matrix_d operator()(const vector_d& shared_params,
+                      const vector_d& job_specific_params,
+                      const std::vector<double>& x_r,
+                      const std::vector<int>& x_i,
+                      std::ostream* msgs = 0) const {
+    const size_type num_shared_params = shared_params.rows();
+    const size_type num_job_specific_params = job_specific_params.rows();
+    const size_type num_params = num_shared_params + num_job_specific_params;
+    matrix_d out(1 + num_params, 0);
+
+    try {
+      start_nested();
+      vector_v shared_params_v(num_shared_params);
```

Comment:
This is suboptimal because it allocates new autodiff variables with the nullary constructor on the stack.  It'd be better to start at size zero, reserve, then use `emplace_back` to construct the elements.  Then no copies are needed.

Given the number of times this loop's called (four times here), it'd make sense to pull out a function like the following:

```
// assumes x1 is size 0 to begin with, otherwise want to resize it to start
template <typename T1, typename T2>
void fill_construct(std::vector<T1>& x1, const std::vector<T2>& x2) {
  x1.reserve(x2.size());
  for (const auto& y : x2)
    x1.emplace_back(y);
}
```

This is also used in the `double`/`var` and `var`/`double` signatures.

Reply:
Or could I use the `to_var` function here? I think this is defined already in stan-math. I just discovered this utility function recently - or is what you propose more efficient?

Reply:
I'm not sure---it'll depend on compilation removing return copies where possible, because `to_var` has a function return.  Now if we had a `to_var` that took the collection to fill, that'd be general enough to use everywhere, but a bit clunky.

Reply:
Correct. I am assuming that the compiler is clever here.

... but wait: What you suggest does not work, because `vector_v` is an Eigen construct. Hence there is no `emplace_back` anyways. Maybe the `to_var` is not so bad then?